libeos-payg/provider-loader.c

```diff
@@ -0,0 +1,310 @@
+
+  if (!peas_engine_load_plugin (self->engine, plugin_info))
+    {
+      g_autoptr(GError) local_error = NULL;
+      gboolean ret = peas_plugin_info_is_available (plugin_info,
+                                                    &local_error);
+      if (ret)
+        {
+          /* Should not happen: peas_engine_load_plugin() should cause the
+           * PeasPluginInfo's error to be set.
+           */
+          g_warning ("Failed to load plugin %s", name);
+        }
+      else
+        {
+          g_warning ("Failed to load plugin %s: %s",
+                     name, local_error->message);
+        }
+
+      return;
+    }
+
+  if (!peas_engine_provides_extension (self->engine,
+                                       plugin_info,
+                                       EPG_TYPE_PROVIDER))
+    return;
+
+  extension = peas_engine_create_extension (self->engine,
+                                            plugin_info,
+                                            EPG_TYPE_PROVIDER,
+                                            NULL);
+  if (extension == NULL)
+    {
+      /* Should not happen: we just checked that plugin_info can do this. */
+      g_warning ("Failed to create EpgProvider from %s", name);
+      return;
+    }
+
+  if (!G_IS_ASYNC_INITABLE (extension))
+    {
+      g_warning ("EpgProvider from %s is not GAsyncInitable", name);
+      return;
+    }
+
+  epg_multi_task_increment (task);
+  g_async_initable_init_async (G_ASYNC_INITABLE (g_steal_pointer (&extension)),
+                               g_task_get_priority (task),
+                               g_task_get_cancellable (task),
+                               provider_init_cb,
+                               g_object_ref (task));
+}
+
+static void
+provider_init_cb (GObject      *source_object,
+                  GAsyncResult *result,
+                  gpointer      user_data)
+{
+  g_autoptr(EpgProvider) provider = EPG_PROVIDER (source_object);
+  g_autoptr(GTask) task = G_TASK (user_data);
+  g_autoptr(GError) error = NULL;
+
+  if (!g_async_initable_init_finish (G_ASYNC_INITABLE (provider),
+                                     result,
+                                     &error))
+    {
+      g_warning ("Failed to initialize %s: %s",
+                 G_OBJECT_TYPE_NAME (provider),
+                 error->message);
+      maybe_return (task, NULL);
+    }
+  else
+    {
+      maybe_return (task, g_steal_pointer (&provider));
+    }
+}
+
+static void
+maybe_return (GTask       *task,
+              EpgProvider *provider)
+{
+  GPtrArray *providers = g_task_get_task_data (task);
+
+  if (provider != NULL)
+    g_ptr_array_add (providers, g_steal_pointer (&provider));
+
+  /* At present, we always return success, even if we found >= 1 plugin but all
+   * failed to load/initialize for some reason. Should we return an error in
+   * that case? If so: if two plugins are found and both fail to load, which
+   * error should we return?
```

Comment:
I guess it depends where you want the policy of “exactly one provider must be loaded” to be implemented. Given that the code to load the internal fallback provider is not in the `ProviderLoader`, that suggests to me that the `ProviderLoader` should just return the array of providers it to loaded (which may be empty), and let the caller figure out what to do next (which might include erroring).

That appears to be what you’ve done anyway. Maybe just turn this comment into an explanation of that.